### PR TITLE
DDF-1809 AttributeQueryClient now uses CXF to send out its requests

### PR DESCRIPTION
**Pull request by roelens8**
-Replaced the HttpUrlConnection with a dispatcher from the Jax-WS Dispatch API (supported by CXF), to send out its request to an external attribute store. There are three additional configuration fields within this feature, and all are required for creating the dynamic service and dispatcher: Location/Address of a WSDL, a Service name from the WSDL, and a Port name that is supported by the service.

-Refactored the unit tests to work with the new dispatcher.

-Restructured project directory into two modules: 
1. A common module that contains all of the code and is packaged as a jar. 
2. An implementation bundle that just contains the blueprint/metatype of the feature and embeds the new common jar.

@coyotesqrl @jckilmer @rzwiefel (@ryeats is the hero) @tbatie @wbarnie 
@stustison @pklinef

### CODE REVIEW

**Comment by review-ninja** [comment]
<a href="https://app.review.ninja/codice/ddf/pull/478" target="_blank">ReviewNinja

File: platform/security/sts/security-sts-attributequeryclaimshandler/src/main/java/org/codice/ddf/security/claims/attributequery/AttributeQueryClient.java [hunk 1]
```diff
@@ -161,50 +159,21 @@ private Element createSoapMessage(AttributeQuery attributeQuery)
...
-        }
-
+    private Assertion retrieveResponse(Document requestDocument) throws AttributeQueryException {
         try {
             Assertion assertion = null;
-            DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory.newInstance();
-            documentBuilderFactory.setNamespaceAware(true);
-            Document responseDocument = documentBuilderFactory.newDocumentBuilder()
-                    .parse(connection.getInputStream());
+            Document responseDocument = (responseDocument = sendRequest(requestDocument)) == null ?
```
Comment (pklinef) [inline platform/security/sts/security-sts-attributequeryclaimshandler/src/main/java/org/codice/ddf/security/claims/attributequery/AttributeQueryClient.java]: This line is very confusing.  I would suggest avoiding variable shadowing.  I also do not see what the ternary is doing.

Comment (roelens8) [inline platform/security/sts/security-sts-attributequeryclaimshandler/src/main/java/org/codice/ddf/security/claims/attributequery/AttributeQueryClient.java]: !fixed

**Comment by stustison** [comment]
I'm curious why we went the route of using the Dispatcher stuff instead of using the SecureProxyServiceFactoryImpl to send the request. If we go with the Dispatcher route, we lose all the nice security stuff that we've set up in the ProxyServiceFactory, and it honestly looks like more work to me. Plus you're forced to sign the SOAP message yourself instead of just having the factory do it for you.

**Comment by roelens8** [comment]
Dispatch API is a better solution as it is more generic and can work with any WSDL.

**Comment by tbatie** [comment]
:+1:

**Comment by stustison** [comment]
:shipit:

**Comment by ryeats** [comment]
:+1:

**Comment by rzwiefel** [comment]
:+1:

**Comment by coyotesqrl** [comment]
:+1:

File: platform/security/sts/security-sts-attributequeryclaimshandler-common/pom.xml [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>security-sts-pom</artifactId>
+        <groupId>ddf.security.sts</groupId>
+        <version>2.9.0-SNAPSHOT</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>security-sts-attributequeryclaimshandler-common</artifactId>
```
Comment (stustison) [inline platform/security/sts/security-sts-attributequeryclaimshandler-common/pom.xml]: This doesn't follow our naming scheme. Use security-sts-attributequerycommon

File: platform/security/sts/security-sts-attributequeryclaimshandler-common/pom.xml [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+    <name>DDF :: Security :: STS :: AttributeQuery Claims Handler :: Common</name>
```
Comment (stustison) [inline platform/security/sts/security-sts-attributequeryclaimshandler-common/pom.xml]: DDF :: Security :: STS :: Attribute Query Common

Comment (roelens8) [inline platform/security/sts/security-sts-attributequeryclaimshandler-common/pom.xml]: Thanks, will fix that.

File: platform/security/sts/security-sts-attributequeryclaimshandler-common/src/main/java/org/codice/ddf/security/claims/attributequery/common/AttributeQueryClaimsHandler.java [hunk 4]
```diff
@@ -226,7 +243,71 @@ protected ProcessedClaim createSingleValuedClaim(String claimType, String claimV
...
+
+    /**
+     * Creates a dynamic service from the provided wsdl location.
+     */
+    protected Service createService() {
+        Service service = null;
+        if (StringUtils.isNotBlank(wsdlLocation) && StringUtils.isNotBlank(serviceName)) {
+            service = Service.create(this.getClass()
+                    .getClassLoader()
+                    .getResource(wsdlLocation), QName.valueOf(serviceName));
```
Comment (stustison) [inline platform/security/sts/security-sts-attributequeryclaimshandler-common/src/main/java/org/codice/ddf/security/claims/attributequery/common/AttributeQueryClaimsHandler.java]: We should not be relying on a WSDL that is compiled with this code. The WSDL will be remote in almost every situation.

File: platform/security/sts/security-sts-attributequeryclaimshandler-common/src/test/java/org/codice/ddf/security/claims/attributequery/common/TestAttributeQueryClient.java [hunk 5]
```diff
@@ -100,13 +63,11 @@ protected HttpsURLConnection createHttpsUrlConnection(URL url) {
 
     private static final String ISSUER = "testIssuer";
 
-    private String username = "admin";
+    private static final String EXTERNAL_ATTRIBUTE_STORE = "https://localhost:8993/";
```
Comment (stustison) [inline platform/security/sts/security-sts-attributequeryclaimshandler-common/src/test/java/org/codice/ddf/security/claims/attributequery/common/TestAttributeQueryClient.java]: Is this going to break when this unit tests runs when a DDF is also running on the system?

Comment (roelens8) [inline platform/security/sts/security-sts-attributequeryclaimshandler-common/src/test/java/org/codice/ddf/security/claims/attributequery/common/TestAttributeQueryClient.java]: It shouldn't, but I will change it to use the SystemBaseUrl anyway.

File: platform/security/sts/security-sts-attributequeryclaimshandler-common/pom.xml (hunk 2 above)
Comment (roelens8) [inline platform/security/sts/security-sts-attributequeryclaimshandler-common/pom.xml]: Changed.

**Comment by ryeats** [comment]
:+1: heroed

**Pull request merged**